frame/contracts/src/storage.rs

```diff
@@ -190,18 +207,103 @@ where
+			Err(Error::<T>::DeletionQueueFull.into())
+		} else {
+			DeletionQueue::append(DeletedContract {
+				pair_count: contract.total_pair_count,
+				trie_id: contract.trie_id.clone(),
+			});
+			Ok(())
+		}
+	}
+
+	/// Calculates the weight that is necessary to remove one key from the trie and how many
+	/// of those keys can be deleted from the deletion queue given the supplied queue length
+	/// and weight limit.
+	pub fn deletion_budget(queue_len: usize, weight_limit: Weight) -> (u64, u32) {
+		let base_weight = T::WeightInfo::on_initialize();
+		let weight_per_queue_item = T::WeightInfo::on_initialize_per_queue_item(1) -
+			T::WeightInfo::on_initialize_per_queue_item(0);
+		let weight_per_key = T::WeightInfo::on_initialize_per_trie_key(1) -
+			T::WeightInfo::on_initialize_per_trie_key(0);
+		let decoding_weight = weight_per_queue_item.saturating_mul(queue_len as Weight);
+
+		// `weight_per_key` being zero makes no sense and would constitute a failure to
+		// benchmark properly. We opt for not removing any keys at all in this case.
+		let key_budget = weight_limit
+			.saturating_sub(base_weight)
+			.saturating_sub(decoding_weight)
+			.checked_div(weight_per_key)
+			.unwrap_or(0) as u32;
+
+		(weight_per_key, key_budget)
+	}
+
+	/// Delete as many items from the deletion queue possible within the supplied weight limit.
+	///
+	/// It returns the amount of weight used for that task or `None` when no weight was used
+	/// apart from the base weight.
+	pub fn process_deletion_queue_batch(weight_limit: Weight) -> Option<Weight> {
+		// We can decode the queue eagerly because we cap the queue depth and assume
+		// that the weight is large enough to do at least one round of deletions.
+		// We can assume that because any other static configuration of the runtime
+		// would make no sense at all.
+		let mut queue = DeletionQueue::get();
+		if queue.is_empty() {
+			return None;
+		}
+
+		let (weight_per_key, mut remaining_key_budget) = Self::deletion_budget(
+			queue.len(),
+			weight_limit,
+		);
+
+		while !queue.is_empty() && remaining_key_budget > 0 {
+			// Cannot panic due to loop condition
+			let trie = &mut queue[0];
+			let pair_count = trie.pair_count;
+			let outcome = child::kill_storage(
+				&crate::child_trie_info(&trie.trie_id),
+				Some(remaining_key_budget),
+			);
+			if pair_count > remaining_key_budget {
+				// Cannot underflow because of the if condition
+				trie.pair_count -= remaining_key_budget;
+			} else {
+				// We do not care to preserve order. The contract is deleted already and
+				// noone waits for the trie to be deleted.
+				let removed = queue.swap_remove(0);
+				match outcome {
+					// This should not happen as our budget was large enough to remove all keys.
+					KillOutcome::SomeRemaining => {
+						debug::warn!(
+							"After deletion keys are remaining in this child trie: {:?}",
+							removed.trie_id,
```

Comment:
At this point, what happens? We need to use governance to remove the trie?

Reply:
This error should not happen. It indicates a programming error somewhere in the contracts module with regard to storage accounting (client and contracts think differently about how many items are in the trie).

So yes, this would need governance to clean up some storage after a potential bug is fixed.

Reply:
Makes sense. Promoted it to an error.